---
title: "Simulating Single-Cell Multi-Omics Data with MOSim"
author: 
  - name: "Carolina Monzó"
  - name: "Ángeles Arzalluz-Luque"
  - name: "Arianna Febbo"
  - name: "Sonia Tarazona"
date: "`r Sys.Date()`"
vignette: >
  %\VignetteIndexEntry(Wiki of how to use scMOSim)
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
output: 
  BiocStyle::html_document
editor_options: 
  markdown: 
    wrap: 72
---

# Introduction

Welcome to the `MOSim` package, a versatile tool for simulating bulk and
single-cell multi-omics data. In this vignette, we will explore how to
create synthetic single-cell data, focusing on single-cell RNA-seq
(scRNA-seq) and single-cell ATAC-seq (scATAC-seq) data. Using the
`MOSim` package, you can generate custom multi-omics datasets for
various experimental conditions, making it an essential resource for
testing and validating analysis methods, or creating benchmark datasets.

# Installation

Before we dive into the exciting world of data simulation, you'll need
to install the `MOSim` package. You can easily obtain it from CRAN using
the following commands:

```{r installing, eval = FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE)) 
    install.packages("BiocManager")

BiocManager::install("MOSim")

# For the latest development version
install.packages("devtools")
devtools::install_github("ConesaLab/MOSim")

```

# Simulating Single-Cell Multi-Omics Data

The core of data simulation lies in the scMOSim function, which allows
you to create synthetic single-cell multi-omics data. Let's explore a
typical example of its usage:

```{r test_run, eval = FALSE}
#library(MOSim)

devtools::load_all()

# Create a list of omics data types (e.g., scRNA-seq and scATAC-seq)
omicsList <- sc_omicData(list("scRNA-seq", "scATAC-seq"))

# Define cell types for your experiment
cell_types <- list(
  'CD4_TEM' = c(1:60),
  'cDC' = c(299:310),
  'Memory_B' = c(497:510),
  'Treg' = c(868:900)
)

# Load an association list containing peak IDs related to gene names
associationList <- MOSim::associationList

# Simulate multi-omics data with specific parameters
testing_groups <- scMOSim(
  omicsList,
  cell_types,
  numberReps = 1,
  numberGroups = 3,
  diffGenes = list(c(0.1, 0.2), c(0.1, 0.2)),
  minFC = 0.25,
  maxFC = 4,
  numberCells = NULL,
  mean = NULL,
  sd = NULL,
  regulatorEffect = list(c(0.1, 0.2), c(0.1, 0.2), c(0.1, 0.2)),
  associationList = associationList
)
```

In the example above, we load omics data types, specify experimental
conditions and cell types, and load an association list. The scMOSim
function lets us simulate multi-omics data with various parameters, such
as the number of replicates, differentially expressed genes, and
regulatory effects.

# Data Preparation

Before diving into simulation, it's essential to have your data ready.
The `sc_omicData` function aids in preparing your data for simulation.
It accepts the following inputs:

-   omics_types: A list of omics data types, which can be "scRNA-seq" or
    "scATAC-seq."

-   data (optional): A user-inputted list of matrices with features as
    rows and cells as columns. If data is NULL, the default data from
    the SeuratData package is used.

## Using the Provided PBMC Dataset

The `MOSim` package offers the option to simulate scRNA and scATAC count
data resembling the multi-omics PBMC dataset from `SeuratData`. This
dataset, subset for computational reasons, contains count data for CD4
TEM, cDC, Memory B, and Treg cells.

## Providing Custom Data

`scMOSim` also allows you to simulate data resembling characteristics of
a dataset of your choice. To do so, you need to format your data using
the `sc_omicData` function. Supported input formats include:

-   Count matrix
-   Seurat array For example, to format your data extracted from the
    original Seurat object, you can follow these steps:

```{r custom data, eval = FALSE}
# This is done to get a dataset to extract a matrix from (for example purposes)
scRNA <- MOSim::sc_omicData("scRNA-seq")
count <- scRNA[["scRNA-seq"]]
options(Seurat.object.assay.version = "v3")
Seurat_obj <- Seurat::CreateAssayObject(counts = count, assay = 'RNA')
omic_list_user <- sc_omicData(c("scRNA-seq"), c(Seurat_obj))
```

The resulting omic_list_user is a named list with "scRNA-seq" as the
name and your count matrix as the value.

# Running the Simulation: scMOSim

## Default scMOSim Simulation

`scMOSim` can simulate scRNA and scATAC count matrices without providing
any additional arguments. For a basic simulation, you only need to input
the omics list and cell types. Here's how it's done:

```{r default, eval = FALSE}
omic_list <- sc_omicData(list("scRNA-seq"))
cell_types <- list('CD4_TEM' = c(1:298), 'cDC' = c(299:496), 'Memory_B' = c(497:867), 
                   'Treg' = c(868:1029))

sim <- scMOSim(omic_list, cell_types)
```

This will result in simulated raw count matrices for scRNA.

## Customizing the scMOSim Simulation

The `scMOSim` function offers a range of parameters to fine-tune your
simulation:

-   omics: A named list specifying the omics data types.
-   cellTypes: A list specifying the cell types.
-   numberReps (optional): The number of biological replicates.
-   numberGroups (optional): The number of different biological groups
    (experimental conditions).
-   diffGenes (optional): To simulate differentially expressed genes
    between groups.
-   minFC (optional): Threshold for downregulated genes.
-   maxFC (optional): Threshold for upregulated genes.
-   numberCells (optional): A vector specifying the number of cells per
    cell type.
-   mean and sd (optional): Vectors for mean and standard deviation of
    expression per cell type.
-   noiseRep (optional): Standard deviation between biological
    replicates.
-   noiseGroup (optional): Standard deviation between groups.
-   feature_no (optional): Total number of features to distribute
    between co-expression clusters.
-   clusters (optional): Number of co-expression patterns to simulate.
-   cluster_size (optional): Number of features in each co-simulation
    cluster. For example, to simulate data with specific settings, you
    can use the following code:

```{r testing, eval = FALSE}
omic_list <- sc_omicData(c("scRNA-seq", "scATAC-seq"))
cell_types <- list('CD4_TEM' = c(1:10), 'cDC' = c(299:305), 'Memory_B' = c(497:510), 'Treg' = 
                     c(868:878))
sim <- scMOSim(omic_list, cell_types, numberReps = 2, 
               numberGroups = 2, diffGenes = list(c(0.2, 0.3)), feature_no = 8000, 
               clusters = 8, mean = c(2*10^6, 1*10^6,2*10^6, 1*10^6), 
               sd = c(5*10^5, 2*10^5, 5*10^5, 2*10^5), 
               regulatorEffect = list(c(0.1, 0.2), c(0.1, 0.2)))
```

# Working with Simulation Results

## The scMOSim Simulation Object

The result of your simulation is stored in a named list with 'sim_sc +
omic name' as names and Seurat objects as values. Each Seurat object
contains the synthetic count matrices for your experiment. Other
relevant information included in the object are:

-   cellTypes: A list specifying the columns in each simulated matrix
    that correspond to each cell type.

-   patterns: Matrix of co-expression patterns affecting the genes
    throughout cell-types.

-   FC: list of Fold Changes applied to each gene to simulate
    differential expression between experimental groups.

-   AssociationMatrices: gene/peak association matrices including
    differential expression and regulatory relationships.

-   Variability: added variability matrices to add dispersion to
    experimental groups and biological replicates.

## Retrieving Simulation Settings

To access simulation settings and other details, you can use the
`scOmicSettings` function. This provides information about the
relationship between genes and peaks, differentially expressed genes,
regulator types, expression patterns, and fold changes for each gene and
peak compared to group 1.

```{r settings, eval = FALSE}
settings <- scOmicSettings(sim)
```

## Accessing the Count Data Matrices

You can extract the simulated matrices for all experimental conditions
and biological replicates using the `scOmicResults` function. This
provides you with the synthetic data for further analysis and
visualization.

```{r results, eval = FALSE}
res <- scOmicResults(sim)
```

# Example full run and visualizations

Let's run a full experiment and plot some interesting facts.

```{r sim, eval = FALSE}
omic_list <- sc_omicData(c("scRNA-seq", "scATAC-seq"))
cell_types <- list('CD4_TEM' = c(1:60), 'cDC' = c(299:310), 'Memory_B' = c(497:520), 'Treg' = 
                     c(868:900))
sim <- scMOSim(omic_list, cell_types, numberCells = c(10, 20, 30, 40), numberReps = 2, 
               numberGroups = 2, diffGenes = list(c(0.2, 0.3)), feature_no = 8000, 
               clusters = 8, mean = c(2*10^6, 1*10^6,2*10^6, 1*10^6), 
               sd = c(5*10^5, 2*10^5, 5*10^5, 2*10^5), 
               regulatorEffect = list(c(0.2, 0.1), c(0.1, 0.3)))
```

Plotting the simulated clusters for scRNAseq data of Replicate sample 1
of Group 2

```{r plot clusters RNA, eval = FALSE}
library(cowplot)
theme_set(theme_cowplot())

cell_types <- sim$cellTypes
# Scale isoforms to show the clustering patterns
coexpr.scaled <- acorde::scale_isoforms(sim$Group_2$Rep_1$`sim_scRNA-seq`$counts, 
                                isoform_col = NULL)
# fillna
coexpr.scaled[is.na(coexpr.scaled)] <- 0

# create cell-to-cell-type ID table
ct <- tibble::tibble(Cell = colnames(sim$Group_2$Rep_1$`sim_scRNA-seq`$counts),
                            cell_type = rep(names(cell_types), times = lengths(cell_types)))

# Me falta una lista de listas con los genes de cada cluster


# compute average-by-cell type cluster patterns
cluster_patterns <- purrr::map(sim$Clusters_list$`Clusters_scRNA-seq`,
                    ~acorde::calculate_cluster_profile(coexpr.scaled,
                                               isoform_ids = .,
                                               id_table = ct,
                                               isoform_col = "transcript"))

pattern_plots <- purrr::map(cluster_patterns,
                     acorde::plot_cluster_profile,
                     ct_labels = c("CD4_TEM", "cDC", "Memory_B", "Treg"))

# Plot the clusters
plot_grid(plotlist = pattern_plots, 
          labels = NULL, 
          ncol = 4)
```

We can do the same with the ATAC data:

```{r plot clusters ATAC, eval = FALSE}
coexpr.scaledATAC <- acorde::scale_isoforms(sim$Group_2$Rep_1$`sim_scATAC-seq`$counts, isoform_col = NULL)

coexpr.scaledATAC[is.na(coexpr.scaledATAC)] <- 0

# create cell-to-cell-type ID table
ctATAC <- tibble::tibble(Cell = colnames(sim$Group_2$Rep_1$`sim_scATAC-seq`$counts),
                            cell_type = rep(names(cell_types), times = lengths(cell_types)))

# compute average-by-cell type cluster patterns
cluster_patternsATAC <- map(sim$Clusters_list$`Clusters_scATAC-seq`,
                    ~acorde::calculate_cluster_profile(coexpr.scaledATAC,
                                               isoform_ids = .,
                                               id_table = ctATAC,
                                               isoform_col = "transcript"))


## Select only the clusters that don't have NAs


pattern_plotsATAC <- map(cluster_patternsATAC,
                     acorde::plot_cluster_profile,
                     ct_labels = c("CD4_TEM", "cDC", "Memory_B", "Treg"))

plot_grid(plotlist = pattern_plotsATAC, 
          labels = NULL, 
          ncol = 4)
```

Back to the simulated scRNA-seq. Plot the number of features in the
sample, and the counts per feature.

```{r plot violins, eval = FALSE}
VlnPlot(sim$Group_2$Rep_1$`sim_scRNA-seq`, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
```

Add metadata to the Seurat objects to be able to plot samples together,
log transform and plot the variable features:

```{r metadata, eval = FALSE}
sim$Group_1$Rep_1$`sim_scRNA-seq` <- Seurat::AddMetaData(sim$Group_1$Rep_1$`sim_scRNA-seq`, list("Group" = rep("Group_1", length(colnames(sim$Group_1$Rep_1$`sim_scRNA-seq`))), "Rep" = rep("Rep_1", length(colnames(sim$Group_1$Rep_1$`sim_scRNA-seq`)))))

sim$Group_1$Rep_2$`sim_scRNA-seq` <- Seurat::AddMetaData(sim$Group_1$Rep_2$`sim_scRNA-seq`, list("Group" = rep("Group_1", length(colnames(sim$Group_1$Rep_2$`sim_scRNA-seq`))), "Rep" = rep("Rep_2", length(colnames(sim$Group_1$Rep_2$`sim_scRNA-seq`)))))

sim$Group_2$Rep_1$`sim_scRNA-seq` <- Seurat::AddMetaData(sim$Group_2$Rep_1$`sim_scRNA-seq`, list("Group" = rep("Group_2", length(colnames(sim$Group_2$Rep_1$`sim_scRNA-seq`))), "Rep" = rep("Rep_1", length(colnames(sim$Group_2$Rep_1$`sim_scRNA-seq`)))))

sim$Group_2$Rep_2$`sim_scRNA-seq` <- Seurat::AddMetaData(sim$Group_2$Rep_2$`sim_scRNA-seq`, list("Group" = rep("Group_2", length(colnames(sim$Group_2$Rep_2$`sim_scRNA-seq`))), "Rep" = rep("Rep_2", length(colnames(sim$Group_2$Rep_2$`sim_scRNA-seq`)))))

group1_1 <- NormalizeData(object = sim$Group_1$Rep_1$`sim_scRNA-seq`, normalization.method = "LogNormalize", scale.factor = 10000)

group1_2 <- NormalizeData(object = sim$Group_1$Rep_2$`sim_scRNA-seq`, normalization.method = "LogNormalize", scale.factor = 10000)

group2_1 <- NormalizeData(object = sim$Group_2$Rep_1$`sim_scRNA-seq`, normalization.method = "LogNormalize", scale.factor = 10000)

group2_2 <- NormalizeData(object = sim$Group_2$Rep_2$`sim_scRNA-seq`, normalization.method = "LogNormalize", scale.factor = 10000)


group1_1 <- FindVariableFeatures(group1_1, selection.method = "vst", nfeatures = 2000)
group1_2 <- FindVariableFeatures(group1_2, selection.method = "vst", nfeatures = 2000)
group2_1 <- FindVariableFeatures(group2_1, selection.method = "vst", nfeatures = 2000)
group2_2 <- FindVariableFeatures(group2_2, selection.method = "vst", nfeatures = 2000)

p1_1 <- LabelPoints(plot = VariableFeaturePlot(group1_1), points = head(VariableFeatures(group1_1), 10), repel = TRUE)
p1_2 <- LabelPoints(plot = VariableFeaturePlot(group1_2), points = head(VariableFeatures(group1_2), 10), repel = TRUE)
p2_1 <- LabelPoints(plot = VariableFeaturePlot(group2_1), points = head(VariableFeatures(group2_1), 10), repel = TRUE)
p2_2 <- LabelPoints(plot = VariableFeaturePlot(group2_2), points = head(VariableFeatures(group2_2), 10), repel = TRUE)

## Plot the variability of the first replicate of both groups
p1_1 + p1_2
```

Make PCAs and elbowplots

```{r make pcas, eval = FALSE}
# PCAS
genesp1_1 <- rownames(group1_1)
group1_1 <- ScaleData(group1_1, features = genesp1_1)
group1_1 <- RunPCA(group1_1, features = VariableFeatures(object = group1_1), npcs = 10)
DimPlot(group1_1, reduction = "pca")

genesp1_2 <- rownames(group1_2)
group1_2 <- ScaleData(group1_2, features = genesp1_2)
group1_2 <- RunPCA(group1_2, features = VariableFeatures(object = group1_2), npcs = 10)

genesp2_1 <- rownames(group2_1)
group2_1 <- ScaleData(group2_1, features = genesp2_1)
group2_1 <- RunPCA(group2_1, features = VariableFeatures(object = group2_1), npcs = 10)

genesp2_2 <- rownames(group2_2)
group2_2 <- ScaleData(group2_2, features = genesp2_2)
group2_2 <- RunPCA(group2_2, features = VariableFeatures(object = group2_2), npcs = 10)

DimPlot(group1_1, reduction = "pca") 
DimPlot(group1_2, reduction = "pca")
DimPlot(group2_1, reduction = "pca")
DimPlot(group2_2, reduction = "pca")

ElbowPlot(group1_1, ndims = 50)
ElbowPlot(group1_2, ndims = 50)
ElbowPlot(group2_1, ndims = 50)
ElbowPlot(group2_1, ndims = 50)
```

Plot also UMAP dimensionality reduction

```{r make umaps, eval = FALSE}
group1_1 <- RunUMAP(group1_1, dims = 1:10)
group1_2 <- RunUMAP(group1_2, dims = 1:10)
group2_1 <- RunUMAP(group2_1, dims = 1:10)
group2_2 <- RunUMAP(group2_2, dims = 1:10)

DimPlot(group1_1, reduction = "umap") 
DimPlot(group1_2, reduction = "umap") 
DimPlot(group2_1, reduction = "umap") 
DimPlot(group2_2, reduction = "umap")
```

Integration of data from replicates and groups using Seurat:

```{r seurat integration, eval = FALSE}
gtogether <- list(g1_rep1 = group1_1, g1_rep2 = group1_2,
           g2_rep1 = group2_1, g2_rep2 = group2_2)

# Normalize and identify variable features for each dataset independently
gtogether <- lapply(X = gtogether, FUN = function(x) {
    x <- NormalizeData(x)
    x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 2000)
})

# Select features that are repeatedly variable across datasets for integration
featurestogether <- SelectIntegrationFeatures(object.list = gtogether)

immune.anchorstogether <- FindIntegrationAnchors(object.list = gtogether, anchor.features = featurestogether)
immune.combinedtogether <- IntegrateData(anchorset = immune.anchorstogether, k.weight = 40)

# specify that we will perform downstream analysis on the corrected data note that the
# original unmodified data still resides in the 'RNA' assay
DefaultAssay(immune.combinedtogether) <- "integrated"

# Run the standard workflow for visualization and clustering
immune.combinedtogether <- ScaleData(immune.combinedtogether, verbose = FALSE)
immune.combinedtogether <- RunPCA(immune.combinedtogether, npcs = 30, verbose = FALSE)
immune.combinedtogether <- RunUMAP(immune.combinedtogether, reduction = "pca", dims = 1:30)
immune.combinedtogether <- FindNeighbors(immune.combinedtogether, reduction = "pca", dims = 1:30)
immune.combinedtogether <- FindClusters(immune.combinedtogether, resolution = 0.5)

# Visualization, plot the clustering of groups and separation by celltypes
p1g1 <- DimPlot(immune.combinedtogether, reduction = "umap", group.by = "Group")
p2g1 <- DimPlot(immune.combinedtogether, reduction = "umap", label = FALSE)
p1g1 + p2g1
```

# How to cite scMOSim

The MOSim package is currently on biorxiv:

Monzó C.\*, Martínez-Mira C\*., Arzalluz-Luque A., Febbo A., Conesa A.\+,
Tarazona S.\+ (2023) MOSim: Multi-layer regulatory network simulator in R.
DOI: 10.1101/421834

<https://www.biorxiv.org/content/10.1101/421834v1>
